gluonnlp/data/word_embedding_training.py

```diff
@@ -0,0 +1,250 @@
+"""Word embedding training.
+
+Datasets and samplers for training of embedding models.
+
+"""
+
+__all__ = ['Text8', 'ContextSampler', 'NegativeSampler']
+
+import os
+import random
+import shutil
+import zipfile
+
+import numpy as np
+from mxnet.gluon.utils import check_sha1, download
+
+from .dataset import CorpusDataset
+from .utils import _get_home_dir
+
+try:
+    from numba import njit, prange
+    numba_njit = njit(nogil=True)
+except ImportError:
+    # Define numba shims
+    prange = range
+
+    def numba_njit(func):
+        return func
+
+
+###############################################################################
+# Datasets
+###############################################################################
+class Text8(CorpusDataset):
+    """Text8 corpus
+
+    http://mattmahoney.net/dc/textdata.html
+
+    Part of the test data for the Large Text Compression Benchmark
+    http://mattmahoney.net/dc/text.html. The first 10**8 bytes of the English
+    Wikipedia dump on Mar. 3, 2006.
+
+    License: https://en.wikipedia.org/wiki/Wikipedia:Copyrights
+
+    Parameters
+    ----------
+    root : str, default '$MXNET_HOME/datasets/text8'
+        Path to temp folder for storing data.
+        MXNET_HOME defaults to '~/.mxnet'.
+
+    """
+
+    archive_file = ('text8.zip', '6c70299b93b7e1f927b42cd8f6ac1a31547c7a2e')
+    data_file = {
+        'train': ('text8', '0dc3edebc970dcc96137e7deda4d9995af9d93de')
+    }
+    url = 'http://mattmahoney.net/dc/'
+
+    def __init__(self, root=os.path.join(_get_home_dir(), 'datasets', 'text8'),
+                 segment='train'):
+        root = os.path.expanduser(root)
+        if not os.path.isdir(root):
+            os.makedirs(root)
+        self._root = root
+        self._segment = segment
+        super(Text8, self).__init__(self._get_data())
+
+    def _get_data(self):
+        archive_file_name, archive_hash = self.archive_file
+        data_file_name, data_hash = self.data_file[self._segment]
+        root = self._root
+        path = os.path.join(root, data_file_name)
+        if not os.path.exists(path) or not check_sha1(path, data_hash):
+            downloaded_file_path = download(self.url + archive_file_name,
+                                            path=root, sha1_hash=archive_hash)
+
+            with zipfile.ZipFile(downloaded_file_path, 'r') as zf:
+                for member in zf.namelist():
+                    filename = os.path.basename(member)
+                    if filename:
+                        dest = os.path.join(root, filename)
+                        with zf.open(member) as source:
+                            with open(dest, 'wb') as target:
+                                shutil.copyfileobj(source, target)
+        return path
+
+
+###############################################################################
+# Samplers
+###############################################################################
+class NegativeSampler(object):
```

Comment:
this should go into sampler.py

Reply:
Done. It has the downside though of requiring specification of a length for the NegativeSampler, given that it does not have an inherent length. (But that's not too bad I guess)